### ISSUE TITLE
TypeError "hover" of undefined, targetId changes during function call caused inside `hover.js` of react-dnd.

### ISSUE THREAD

**Issue by Braweria** [react-dnd/react-dnd#3208]
**Describe the bug**
```bash
Unhandled Runtime Error
TypeError: Cannot read property 'hover' of undefined
```  
The targeted changes during the function call.
Here is a small gif demonstration, below it I will have a written description.  
![react-dnd-issue](https://user-images.githubusercontent.com/62669899/118769173-2b7eff80-b880-11eb-8836-01bc6e4544e4.gif)  
And a gif on what happens on the frontend:  
![react-dnd-issue-frontend](https://user-images.githubusercontent.com/62669899/118769907-f626e180-b880-11eb-9b64-8cf8f114842b.gif)

The gist is, when I want to move a task to an empty board, the TypeError about the hover appears. On further inspection, I found out, that it is caused inside of a function inside the `hover.js` file.  
```js
function hoverAllTargets(targetIds, monitor, registry) {
  // Finally call hover on all matching targets.
  targetIds.forEach(function (targetId) {
    var target = registry.getTarget(targetId);
    target.hover(monitor, targetId);
  });
}
```
On the first run-through, the `targetId` is `T20` and the target can be found. Then there is *another* run and the `targetId` changes to a different ID, it is always another ID on each debugging session. Once the `targetId` is a different ID, it cannot find `target` anymore and is set to `undefined`.  

```js
const [, dropRef] = useDrop({
    accept: ItemTypes.TASK,
    hover: (_, monitor) => {
      if (!hasTasks) {
        dispatchTaskList({
          type: 'moveTask',
          args: [monitor.getItem().taskId, 0, boardIndex],
        });
      }
    },
  });
```
This is the function that causes it, it is inside of my `Board.js`. It is the same function call, to move tasks from board to board when they're already filled. The difference here is, that I have the index (the second argument) set to 0 since the new task will be the first task on that board. I am also checking here if the board has tasks or not.  
This is what the function call looks like on `Task.js`  
```js
  const [, dropRef] = useDrop({
    accept: ItemTypes.TASK,
    hover: (_, monitor) => {
      dispatchTaskList({
        type: 'moveTask',
        args: [monitor.getItem().taskId, taskIndex, boardId],
      });
    },
  });
```

I also don't fully understand what is wrong. I have two examples below, one that works and one that is similar to mine. Both use the same function.

**Reproduction**

https://codesandbox.io/s/react-dnd-kanban-board-with-issue-ek0xf?file=/src/Kanban.js  
Here is my attempt at reproducing the issue. I copied everything 1:1 but removed the MUI styling.

1. Move tasks from one board to another until that board is empty
2. try to move a task into that empty board
3. error

You might have to move the task a few times into an empty board before the error appears. I have also created a local version to check against: https://github.com/Braweria/react-dnd-hover-issue  
I am using NextJS for my project, which the local version also uses. But again, with no styling.

**Expected behavior**
And then there is *this* version: https://codesandbox.io/s/react-dnd-kanban-board-oo3ui?file=/src/App.js  
This is the version that works the way I want it to work. You can move a task into an empty board and it moves the task there and displays it as well. You can do it on every board, no error pops up.

**Desktop (please complete the following information):**

- OS: Win10
- Browser Chrome
- Versions:
```js
    "react-dnd": "^14.0.2",
    "react-dnd-html5-backend": "^14.0.0",
```

**Additional context**
I have double-checked with the working example already and I just cannot wrap my head around, why this doesn't work. What is wrong? I have used my working example as guidance for my project, but now the hover error happens.  
Here is another issue about the hover error, but with no answers: https://github.com/react-dnd/react-dnd/issues/2563  
Here is another working example, but the code is slightly different because it uses nested lists. The similarity lies in that both use `hover` to call functions and change the state. https://codesandbox.io/s/react-dnd-nested-list-example-fehzw

**Comment by Braweria** [comment]
# I managed to find a fix
So after many attempts at trying to fix it myself, I *finally* managed to do it. There are two things that were wrong with my code. One of them is not present in the sandbox example.  
The first thing that was wrong, was my function. I accidentally passed the wrong argument.  
```js
dispatchTaskList({
  type: "moveTask",
  args: [monitor.getItem().taskId, 0, boardIndex]
});
```
This is what I wrote, but the args were supposed to be this `args: [monitor.getItem().taskId, 0, boardId]` So either I just wrote it wrong or autocomplete messed with me.  
Now, with this fix, I managed to move the task to an empty board, but the hover error still occurred.  

Now onto the fix that removed the error. In my project, I am using Material UI, which I did not apply to my examples. Whcih was a mistake, because that was the key to fixing it. In my `Task.js` my return statement was as follows:  
You can see the Paper is wrapped in one div-container.  
```js
dragRef(dropRef(ref));

 return (
    <div ref={previewRef} style={isSpacer ? { height: '100%' } : null}>
        <Paper
          className={classes.task}
          style={isSpacer ? { backgroundColor: 'red', height: '100%' } : null}
          ref={ref}
        >
          <p>{task?.content}</p>
        </Paper>
    </div>
  );
```  
While debugging I remembered seeing in the React Dev Tools, that the Paper component had some `forwardRefs`, so that made me think that the ref I am applying to my paper, is not wrapping my paper, but actually it is drilling into it. So I removed that ref and created another div wrapping that paper with the ref. So now the ref is wrapping my Paper.
```js
  return (
    <div ref={previewRef} style={isSpacer ? { height: '100%' } : null}>
      <div ref={ref}>
        <Paper
          className={classes.task}
          style={isSpacer ? { backgroundColor: 'red', height: '100%' } : null}
        >
          <p>{task?.content}</p>
        </Paper>
      </div>
    </div>
  );
```  

And no more `'hover' of undefined`.  

## **TLDR:** Have the ref on a DOM Element, like a `<div>` and not on a component, like `<Paper>`.
